core/common/src/main/java/alluxio/extensions/ExtensionFactoryRegistry.java

```diff
@@ -201,6 +200,7 @@ private void scan(List<File> files, List<T> factories) {
           LOG.debug("Discovered a factory implementation {} - {} in jar {}", factory.getClass(),
               factory, jarPath);
           register(factory, factories);
+          register(factory);
```

Comment:
```
  /**
   * The base list of factories, which does not include any lib or extension factories. The only
   * factories in the base list will be built-in factories, and any additional factories
   * registered by tests. All other factories will be discovered and service loaded when extension
   * creation occurs.
   */
  private final List<T> mFactories = new CopyOnWriteArrayList<>();
```
we modify the basic assumption of what `mFactories` represents

will it be better to include a new list call `mDynamicLoadedFactories`?

Reply:
1. If mFactories contains -> return
2. If mDynamicLoadedFactories contains -> return
3. If not, renew the mDynamicLoadedFactories?

Reply:
@apc999 What's your thoughts

Reply:
> 1. If mFactories contains -> return
> 2. If mDynamicLoadedFactories contains -> return
> 3. If not, renew the mDynamicLoadedFactories?

Yea, i think it sounds better! By the way, may be that we can add a method "boolean supportsDynamicLoad()" in the public interface ExtensionFactory to distinguish the underFileSystemFactories, likes  the existed method "boolean supportsPath(String path, S conf)"，so we can still only use mFactories. 

What's your thoughts?   @apc999 @LuQQiu @maobaolong

Reply:
> I prefer this way, but is it complicated to add the logic?
Based on your idea, the steps are :  
          1. If mFactories contains &&  !supportsDynamicLoad()     -> return
          2. If mDynamicLoadedFactories contains && !supportsDynamicLoad()    -> return
          3. If not, renew the mDynamicLoadedFactories?

Reply:
The workload SGTM, @apc999 @ggezer @yuzhu PTAL